spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/expressions/TransformExpressions.scala

```diff
@@ -119,3 +118,38 @@ case class IcebergBucketTransform(numBuckets: Int, child: Expression) extends Ic
 
   override def dataType: DataType = IntegerType
 }
+
+case class IcebergTruncateTransform(child: Expression, width: Int) extends IcebergTransformExpression {
+
+  override def children: Seq[Expression] = child :: Nil
+
+  @transient lazy val truncateFunc: Any => Any = child.dataType match {
+    case _: DecimalType =>
+      val t = Transforms.truncate[Any](icebergInputType, width)
+      d: Any =>  Decimal.apply(t(d.asInstanceOf[Decimal].toJavaBigDecimal).asInstanceOf[java.math.BigDecimal])
+    case _: StringType =>
+      val t = Transforms.truncate[String](Types.StringType.get(), width)
+      s: Any => UTF8String.fromString(t(s.toString))
```

Comment:
`TruncateString` accepts a `CharSequence`, not just `String`. It's unfortunate that `UTF8String` doesn't implement `CharSequence`, but we can fairly easily get one from the underlying bytes: `StandardCharsets.UTF_8.decode(ByteBuffer.wrap(s.asInstanceOf[UTF8String].getBytes))`

That will return a `CharSequence` that is not a copy of the bytes, then we can convert that back to a byte buffer and create the `UTF8String` from that.